spec/p2p/v0.34/addressbook.md

```diff
@@ -0,0 +1,368 @@
+  buckets in which there is an address with the same node ID.
+  So, a new address for a node ID which is already present in one bucket is
+  added with 50% of probability; if the node ID is present in two buckets, the
+  probability decreases to 25%; and if it is present in three buckets, the
+  probability is 12.5%.
+
+Addresses are stored in [buckets](#buckets).
+The bucket to which a new address is added is computed based on the address
+itself and its source address.
+An address is not added to a bucket if it is already present there.
+This should not happen with addresses with new node IDs, but it can happen with
+addresses with known node IDs.
+In fact, as the mapping of buckets for new addresses is mainly determined by
+the network group of the source address, the same address can be present in
+more than one bucket provided that it was added with different source addresses.
+
+The new address is also added to the `addrLookup` table, which stores
+`knownAddress` entries indexed by their node IDs.
+If the new address is from an unknown peer, a new entry is added to the
+`addrLookup` table; otherwise, the existing entry is updated with the new
+address.
+Entries of this table contain, among other fields, the list of buckets where
+addresses of a peer are stored.
+The `addrLookup` table is used by most of the address book methods (e.g.,
+`HasAddress`, `IsGood`, `MarkGood`, `MarkAttempt`), as it provides fast access
+to addresses.
+
+### Errors
+
+- if the added address or the associated source address are nil
+- if the added address is invalid
+- if the added address is a node address
+- if the added address ID is of a banned peer
+- if either the added address or the associated source address IDs are configured as private IDs
+- if `routabilityStrict` is set and the address is not routable
+- in case of failures computing the bucket for the new address (`calcNewBucket` method)
+- if the added address instance, which is a new address, is configured as an
+  old address (sanity check of `addToNewBucket` method)
+
+## Need for Addresses
+
+The `NeedMoreAddrs` method verifies whether the address book needs more addresses.
+
+It is invoked by the PEX reactor to define whether to request peer addresses
+to a new outbound peer or to a randomly selected connected peer.
+
+The address book needs more addresses when it has less than `1000` addresses
+registered, counting all buckets for new and old addresses.
+
+## Pick address
+
+The `PickAddress` method returns an address stored in the address book, chosen
+at random with a configurable bias toward new addresses.
+
+It is invoked by the PEX reactor to obtain a peer address to dial, as part of
+its `ensurePeers` routine.
+The bias starts from 10%, when the peer has no outbound peers, increasing by
+10% for each outbound peer the node has, up to 90%, when the node has at least
+8 outbound peers.
+
+The configured bias is a parameter that influences the probability of choosing
```

Comment:
The same comment as above; too low level and not clear what  this logic is trying to achieve.